### PR TITLE
Add a UT for the entire training loop

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/pretrain/test_pretrain_qwen3_next.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import os
+import pytest
+import subprocess
+
+
+def get_bash_cmd(model_args, env_args = None, num_gpus_per_node=8):
```
Comment (Skylion007) [inline tests/unit_tests/pretrain/test_pretrain_qwen3_next.py]: FYI< this code snippit might be useful to you: https://github.com/facebookresearch/habitat-sim/blob/6c26c4f9ae10ad7f534aa9e59a84dff00782741e/tests/helpers/utils.py#L1

File: tests/unit_tests/pretrain/test_pretrain_qwen3_next.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    args = [
+        "python3",
+        "-m",
+        "torch.distributed.launch",
+        "--nproc-per-node=" + str(num_gpus_per_node),
+    ]
+    megatron_path = os.getenv("MEGATRON_PATH", "/opt/Megatron-LM")
```
Comment (Skylion007) [inline tests/unit_tests/pretrain/test_pretrain_qwen3_next.py]: Any reason not use the git directory as the root?

File: tests/unit_tests/pretrain/test_pretrain_qwen3_next.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            **parallel_args,
+            "--save": save_path,
+            "--load": None,  # train from scratch
+            "--exit-interval": 5,
+        },
+        env_args=ENV_ARGS,
+        num_gpus_per_node=8,
+    )
+    # subprocess.run(cmd, env=ENV_ARGS, check=True)
+    ret = os.system(" ".join(cmd))
```
Comment (Skylion007) [inline tests/unit_tests/pretrain/test_pretrain_qwen3_next.py]: Um, this is deprecated for good reason related to potentially invalid arg parsing, please use the newer subprocess run command with the args list or if you must do this:
```suggestion
    ret = os.system(shlex.join(cmd))
```
is far safer

File: tests/unit_tests/pretrain/test_pretrain_qwen3_next.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            **MODELS[model_name],
+            **TRAINING_ARGS,
+            "--save": save_path,
+            "--load": save_path,  # load from checkpoint
+            "--exit-interval": 10,
+        },
+        env_args=ENV_ARGS,
+        num_gpus_per_node=8,
+    )
+    ret = os.system(" ".join(cmd))
```
Comment (Skylion007) [inline tests/unit_tests/pretrain/test_pretrain_qwen3_next.py]:
```suggestion
    ret = os.system(shlex.join(cmd))

File: tests/unit_tests/pretrain/test_pretrain_qwen3_next.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            **MODELS[model_name],
+            **TRAINING_ARGS,
+            **parallel_args,
+            "--save": save_path,
+            "--load": save_path,  # load from checkpoint
+        },
+        env_args=ENV_ARGS,
+        num_gpus_per_node=8,
+    )
+    ret = os.system(" ".join(cmd))
```
Comment (Skylion007) [inline tests/unit_tests/pretrain/test_pretrain_qwen3_next.py]:
```suggestion
    ret = os.system(shlex.join(cmd))

File: tests/unit_tests/pretrain/test_pretrain_qwen3_next.py (hunk 3 above)
Comment (yuzhongw-nvidia) [reply tests/unit_tests/pretrain/test_pretrain_qwen3_next.py]: I want to use `subprocess.run` command, but I failed to specify the environs by the `env` argument and got the following error

```
File "/usr/local/lib/python3.12/dist-packages/torch/distributed/elastic/agent/server/api.py", line 101, in __post_init__
    assert self.entrypoint
           ^^^^^^^^^^^^^^^
AssertionError
```

I'm not sure whether torchrun does not co-work well with `subprocess.run(env=...)`. Do you have any ideas about it? Or I may use `ret = os.system(shlex.join(cmd))` instead. Thanks!

**Comment by yuzhongw-nvidia** [comment]
> Whats the difference between this and a functional test?

I think a functional test may focus more on the numeric, but a UT cares more about the functionality. We want to ensure the corresponding features are runnable, so we would like to add one (or several) UT(s) to cover them. Do I understand correctly? Please correct me if there is anything wrong. Thanks!

**Comment by ko3n1g** [comment]
> > Whats the difference between this and a functional test?
> 
> I think a functional test may focus more on the numeric, but a UT cares more about the functionality. We want to ensure the corresponding features are runnable, so we would like to add one (or several) UT(s) to cover them. Do I understand correctly? Please correct me if there is anything wrong. Thanks!

Functional tests have a mode (`ENABLE_LIGHTWEIGHT`) which only trains for 4 steps and doesnt test for numerics. I think that's what this test here is doing. I'm worried that we're bloating up the unit test suite by adding tests like this. A unit test is meant to be fast and isolated, the training loop is by definition an integration

**Comment by yuzhongw-nvidia** [comment]
> Functional tests have a mode (`ENABLE_LIGHTWEIGHT`) which only trains for 4 steps and doesnt test for numerics. I think that's what this test here is doing. I'm worried that we're bloating up the unit test suite by adding tests like this. A unit test is meant to be fast and isolated, the training loop is by definition an integration

Thanks for your suggestion. Let's move on by re-enabling the qwen3-next's functional test in the `ENABLE_LIGHTWEIGHT` mode, and we may re-open this discussion in the future if necessary.
